### DISCUSSION TITLE
Basics not working - motor forever

### DISCUSSION THREAD

**Discussion by Apollo22251** [pybricks/support#1955] [General]
Hi, all - Happy Thanksgiving for my fellow folks in the US! 

My daughter and I are trying to learn how to use multitasking and/or async (i.e., not drivebase) motor functions to create a MyTask for line squaring in FLL. 

At a very elementary level, I cannot get a single motor to run forever. Please see attached program — it's only one block long, and two blocks in the setup. LMotor won't turn.  It should though, yes?  It's a true 'forever,' just start it up and away it goes? 

Thank you -

<img width="628" alt="Screenshot 2024-11-28 at 2 11 34 PM" src="https://github.com/user-attachments/assets/9e66b9b2-f99e-44bf-9656-dd165c9e5c45">

**Comment by BertLindeman** [comment]
Hello,

**The** problem is that your program starts the motor and is **ready** after that, so the motor does not get the time to run.
You probably want to run the motor until some event occurs. 

But at least you might add a wait block, so you can see that the motor indeed runs.

Hope I state this clear enough.... 
Bert

**Reply by Apollo22251** [reply]
I added a 5000ms wait block after it, and it runs. Thank you. 

Now - I want to interrupt that 5000ms runtime when the sensor sees <20% reflectivity (black on our field). How do I perform that interruption? 

Thank you for your help.

**Reply by BertLindeman** [reply]
Without trying myself! 

All in a while block, maybe ```while True``` so it runs on "forever"
In that loop:
A compare-block with the sensor reflection value to compare against 20
if that is true a motor block to stop it.

Leaving it to you to find the correct blocks 😄

**Reply by Apollo22251** [reply]
Starting here - and it works.... When I place the robot such that a sensor sees a white piece of paper, the motors stop. 

<img width="972" alt="Screenshot 2024-11-28 at 4 20 47 PM" src="https://github.com/user-attachments/assets/95a39d4c-5589-44f4-8b66-e47b076d2adf">

Now, to figure out how to "if ColorL sees white, turn on R Motor until ColorR sees white...." 

Would gladly accept nudges in the right direction.  Many thanks to this community!

**Reply by BertLindeman** [reply]
😄 Nudges you want the motor to give so the robot goes onto the straight and narrow (so to speak).

You learn fast.

In fact is what you write in
> "if ColorL sees white, turn on R Motor until ColorR sees white...."

Already coding, just not yet in blocks or python. 
This is called pseudo code.

You took a multitask block and have two tasks in it and want them both be ready before going on.
There are two motor blocks. 
Replace them both by blocks that do what you wrote in your pseudo code.
Maybe place the two "start the motor" blocks before the multitask block. 
Needed only once until the blocks in the multitask blocks  want the motors to do differently. 

How about such a nudge?

**Reply by laurensvalk** [reply]
@Apollo22251,

A more common (and perhaps simpler) pattern might be as follows:

- Set up a drive base
- Tell it to drive forever (rather then each motor separately)
- Use a wait block to wait until the sensors are triggered. Much like the green logic block you already have, but without the need for loops or multitasking.

**Comment by BertLindeman** [comment]
Time to close for the night here.

Laurens showed  an example of line squaring with one color sensor.

Not what you are aiming at, but it can be useful to look at it.
See [discussion 1699 Line squaring with blocks](https://github.com/orgs/pybricks/discussions/1699)
Also for a parent supporting a daughter.

Interested how you go.

Bert
